socketio/src/client/reconnect.rs

```diff
@@ -0,0 +1,216 @@
+#![allow(unused)]
+use std::{
+    ops::Deref,
+    sync::{Arc, Mutex, RwLock},
+    time::Duration,
+};
+
+pub use super::super::{event::Event, payload::Payload};
+use super::{Client, ClientBuilder};
+use crate::{error::Result, packet::Packet, Error};
+use backoff::ExponentialBackoff;
+use backoff::{backoff::Backoff, ExponentialBackoffBuilder};
+
+#[derive(Clone)]
+pub struct ReconnectClient {
+    builder: ClientBuilder,
+    client: Arc<RwLock<super::Client>>,
+    backoff: ExponentialBackoff,
+    max_reconnect_attempts: u8,
+}
+
+impl ReconnectClient {
+    pub(crate) fn new(builder: ClientBuilder) -> Result<Self> {
+        let builder_clone = builder.clone();
+        let client = builder_clone.connect_manual()?;
+        let backoff = ExponentialBackoffBuilder::new()
+            .with_initial_interval(Duration::from_millis(builder.reconnect_delay))
+            .with_max_interval(Duration::from_millis(builder.reconnect_delay_max))
+            .build();
+        let max_reconnect_attempts = builder.reconnect_attempts;
+
+        let s = Self {
+            builder,
+            client: Arc::new(RwLock::new(client)),
+            backoff: ExponentialBackoff::default(),
+            max_reconnect_attempts,
+        };
+        s.poll_callback();
+
+        Ok(s)
+    }
+
+    pub fn emit<E, D>(&self, event: E, data: D) -> Result<()>
+    where
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO: like js client, buffer emit, resend after reconnect
+        client.emit(event, data)
+    }
+
+    pub fn emit_with_ack<F, E, D>(
+        &self,
+        event: E,
+        data: D,
+        timeout: Duration,
+        callback: F,
+    ) -> Result<()>
+    where
+        F: for<'a> FnMut(Payload, Client) + 'static + Sync + Send,
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO: like js client, buffer emit, resend after reconnect
+        client.emit_with_ack(event, data, timeout, callback)
+    }
+
+    pub fn disconnect(&self) -> Result<()> {
+        let client = self.client.read()?;
+        client.disconnect()
+    }
+
+    fn reconnect(&mut self) -> Result<()> {
+        let mut reconnect_attempts = 0;
+        if self.builder.reconnect {
+            loop {
+                if self.max_reconnect_attempts != 0 // 0 means infinity
+                    && reconnect_attempts > self.max_reconnect_attempts
+                {
+                    break;
+                }
+                reconnect_attempts += 1;
+
+                if let Some(backoff) = self.backoff.next_backoff() {
+                    std::thread::sleep(backoff);
+                }
+
+                if self.do_reconnect().is_err() {
+                    self.do_reconnect();
+                }
```

Comment:
I don't really understand why we need to do this again if it fails the first time, can you elaborate?

Reply:
The right one should be this:
```rust 
if self.do_reconnect().is_ok() {
      break;
}
```
I'll fix it.